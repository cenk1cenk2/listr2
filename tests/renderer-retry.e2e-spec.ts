import delay from 'delay'
import { mockProcessExit, mockProcessStderr, mockProcessStdout } from 'jest-mock-process'

import { Listr } from '@root/index'

describe('show retry', () => {
  let mockExit: jest.SpyInstance<never, [number?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStdout: jest.SpyInstance<boolean, [string, string?, Function?]>

  // eslint-disable-next-line @typescript-eslint/ban-types
  let mockStderr: jest.SpyInstance<boolean, [string, string?, Function?]>

  process.stdout.isTTY = true

  beforeEach(async () => {
    mockExit = mockProcessExit()
    mockStdout = mockProcessStdout()
    mockStderr = mockProcessStderr()
  })

  afterEach(async () => {
    mockExit.mockRestore()
    mockStdout.mockRestore()
    mockStderr.mockRestore()
    jest.clearAllMocks()
  })

  // R34CKurUXSpq65S8ebD6jpUrwmIiYrKa
  it('should retry the main task if any of the subtasks fail', async () => {
    let err: Error | undefined
    try {
      await new Listr(
        [
          {
            title: 'Some type errors',
            task: async (_, task): Promise<void> => {
              await delay(10)
              task.output = 'test'

              await delay(10)
              const retry = task.isRetrying()!
              if (retry.count > 0) {
                task.output = `I am self aware that I am retrying for the ${retry.count}th time.`
              }

              await delay(10)
              throw new Error('This type can not be assigned to type with, oh noes')
            },
            retry: 3
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: { lazy: true }
        }
      ).run()
    } catch (e) {
      err = e
    }

    expect(err).toBeTruthy()
    expect(mockStdout.mock.calls).toMatchSnapshot('R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-out')
    expect(mockStderr.mock.calls).toMatchSnapshot('R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-err')
    expect(mockExit.mock.calls).toMatchSnapshot('R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-exit')
  })

  it('should stop retrying if the task succeeds afterwards', async () => {
    let err: Error | undefined
    try {
      await new Listr(
        [
          {
            title: 'Some type errors',
            task: async (_, task): Promise<void> => {
              const retry = task.isRetrying()

              if (retry?.count === 3) {
                task.title = 'Successed at 3th try.'
              } else {
                throw Error('not enough')
              }
            },
            retry: 3
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          rendererOptions: { lazy: true }
        }
      ).run()
    } catch (e) {
      err = e
    }

    expect(err).toBeFalsy()
    expect(mockStdout.mock.calls).toMatchSnapshot()
    expect(mockStderr.mock.calls).toMatchSnapshot()
    expect(mockExit.mock.calls).toMatchSnapshot()
  })
})
