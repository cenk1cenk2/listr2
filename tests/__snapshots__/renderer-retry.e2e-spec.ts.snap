// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-err 1`] = `
Array [
  Array [
    "[?25l",
  ],
  Array [
    "[?25h",
  ],
]
`;

exports[`show retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-exit 1`] = `Array []`;

exports[`show retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-out 1`] = `
Array [
  Array [
    "[?25l",
  ],
  Array [
    "‚óº Some type errors
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚óº Some type errors
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ùØ Some type errors
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ùØ Some type errors
  ‚Ä∫ test
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ùØ Some type errors
  ‚Ä∫ test
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ùØ Some type errors
  ‚Ä∫ test
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ùØ Some type errors
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ùØ Some type errors
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-1]
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-1]
  ‚Ä∫ test
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-1]
  ‚Ä∫ I am self aware that I am retrying for the 1th time.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-2]
  ‚Ä∫ I am self aware that I am retrying for the 1th time.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-2]
  ‚Ä∫ I am self aware that I am retrying for the 1th time.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-2]
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-2]
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-2]
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-2]
  ‚Ä∫ test
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-2]
  ‚Ä∫ I am self aware that I am retrying for the 2th time.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-3]
  ‚Ä∫ I am self aware that I am retrying for the 2th time.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-3]
  ‚Ä∫ I am self aware that I am retrying for the 2th time.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-3]
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-3]
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-3]
",
  ],
  Array [
    "[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-3]
  ‚Ä∫ test
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚ö† Some type errors [RETRYING-3]
  ‚Ä∫ I am self aware that I am retrying for the 3th time.
",
  ],
  Array [
    "[2K[1A[2K[1A[2K[G",
  ],
  Array [
    "‚úñ This type can not be assigned to type with, oh noes
",
  ],
  Array [
    "[?25h",
  ],
]
`;

exports[`show retry should stop retrying if the task succeeds afterwards 1`] = `Array []`;

exports[`show retry should stop retrying if the task succeeds afterwards 2`] = `Array []`;

exports[`show retry should stop retrying if the task succeeds afterwards 3`] = `Array []`;
